### ISSUE TITLE
(13P*) Full Site Editing: Implement Gutenberg Tracks Events

### ISSUE THREAD

**Issue by apeatling** [Automattic/wp-calypso#33680]
We need to implement better stats tracking inside of Gutenberg. [There is some prior art](https://github.com/Automattic/wp-calypso/pull/33322), added via Jetpack. However, I think we should be adding further event tracking to our Full Site Editing plugin until it's clearer what the future path is into Jetpack.

Here is what I think is an exhaustive list of events we can track. By tracking to this level, we can paint a really clear picture of the journey of a user through an editing session. I believe that's going to be really important to see exactly where they succeed and fail with our full site editing experience.

To be able to see a user's journey through the editor, we need to pass a username or user ID along with each event.

# Update (15th July 2019)

As per `paYJgx-5g-p2#comment-87` the new reduced set of events is as follows (note these have also been renamed to follow the Tracks naming schema - see `PCYsg-4S2-p2`):

- `gutenberg_headerbar_selected_back`
- `gutenberg_headerbar_performed_undo`
- `gutenberg_blockpicker_selected_block`
- `gutenberg_blockoptions_deleted_block`
- `gutenberg_blockarrangement_moved_block_up`
- `gutenberg_blockarrangement_moved_block_down`
- `gutenberg_blockarrangement_dragged_block`

------------

### **Crashes**

I don't know how to reproduce this (except for [these](https://en.forums.wordpress.com/topic/the-editor-has-encountered-an-unexpected-error-4/) [threads](https://github.com/Automattic/wp-calypso/issues/32120), but we definitely need to track it:

<img width="991" alt="Screen Shot 2019-06-05 at 3 00 55 PM" src="https://user-images.githubusercontent.com/1464705/58993595-cf119b00-87a2-11e9-8013-6c6f3f17621a.png">

- [ ] gutenberg_errormessage_encountered_error
- [ ] gutenberg_errormessage_selected_attempt_recovery
- [ ] gutenberg_errormessage_selected_copy_post_text
- [ ] gutenberg_errormessage_selected_copy_error

------------

### **Top Header Bar**

<img width="1279" alt="Screen Shot 2019-06-05 at 1 33 05 PM" src="https://user-images.githubusercontent.com/1464705/58988403-789e5f80-8796-11e9-89c4-aa0f9444b7c6.png">

From left to right:

- [x] gutenberg_headerbar_selected_back
- [x] gutenberg_headerbar_performed_undo
- [x] gutenberg_headerbar_performed_redo
- [ ] gutenberg_headerbar_viewed_content_structure
- [ ] gutenberg_headerbar_viewed_block_navigation
- [ ] gutenberg_headerbar_viewed_preview
- [ ] gutenberg_headerbar_toggled_settings_sidebar
- [ ] gutenberg_headerbar_toggled_sidebar_extension (pass extension name as prop)
- [ ] gutenberg_headerbar_viewed_tools_options_menu

------------

### **Block Picker**

<img width="499" alt="Screen Shot 2019-06-05 at 1 41 19 PM" src="https://user-images.githubusercontent.com/1464705/58988940-a59f4200-8797-11e9-80fb-7b034497a566.png">

- [ ] gutenberg_blockpicker_opened (pass location as prop e.g. headerbar, or ineditor)
- [ ] gutenberg_blockpicker_toggled_blocks_panel (pass panel name as prop)
- [x] gutenberg_blockpicker_selected_block (pass provider and block name concatenated as prop)
- [ ] gutenberg_blockpicker_selected_manage_reusable_blocks

The following two events already exist:

- [ ] wpcom_block_picker_search_term (already exists see: https://github.com/Automattic/wp-calypso/pull/33322)
- [ ] wpcom_block_picker_no_results (already exists see: https://github.com/Automattic/wp-calypso/pull/33322)

I would suggest renaming these to match and including them in the plugin instead. Renaming to "gutenberg_blockpicker_searched_for_block", and "gutenberg_blockpicker_no_search_results".

------------

### **Shortcuts Buttons**

<img width="210" alt="Screen Shot 2019-06-05 at 1 57 45 PM" src="https://user-images.githubusercontent.com/1464705/58990977-0fb9e600-879c-11e9-8538-bb5ea1850947.png">

- [ ] gutenberg_shortcuts_selected_block_shortcut (pass the provider and block name concatenated as prop)

------------

### **Block Options Menu(s)**

<img width="340" alt="Screen Shot 2019-06-05 at 2 06 56 PM" src="https://user-images.githubusercontent.com/1464705/58990645-3b889c00-879b-11e9-8875-8f5488668b63.png">

<img width="474" alt="Screen Shot 2019-06-05 at 2 07 03 PM" src="https://user-images.githubusercontent.com/1464705/58990646-3b889c00-879b-11e9-98d6-72dc802f647a.png">

_We should pass the provider and block name concatenated as a prop to all of these events._

Transform to menu:
- [ ] gutenberg_blockoptions_toggled_transformto_panel
- [ ] gutenberg_blockoptions_performed_transform (pass to and from provider and block name concatenated props)

Menu buttons:
- [ ] gutenberg_blockoptions_selected_option (pass the option name as second prop)

More options menu:
- [ ] gutenberg_blockoptions_selected_show_block_settings
- [ ] gutenberg_blockoptions_duplicated_block
- [ ] gutenberg_blockoptions_inserted_block_before
- [ ] gutenberg_blockoptions_inserted_block_after
- [ ] gutenberg_blockoptions_edited_as_html
- [ ] gutenberg_blockoptions_added_to_reusable_blocks

------------

### **Block Arrangement**

<img width="42" alt="Screen Shot 2019-06-05 at 2 02 58 PM" src="https://user-images.githubusercontent.com/1464705/58990390-a7b6d000-879a-11e9-96d8-e0c874d5e8e9.png">

_We should pass the provider and block name concatenated as a prop to all of these events._

- [x] gutenberg_blockarrangement_moved_block_up
- [x] gutenberg_blockarrangement_moved_block_down
- [x] gutenberg_blockarrangement_dragged_block

------------

### **Settings Sidebar Header**

<img width="147" alt="Screen Shot 2019-06-05 at 2 18 05 PM" src="https://user-images.githubusercontent.com/1464705/58991295-c7e78e80-879c-11e9-9cb0-2512fcc0cd05.png">

- [ ] gutenberg_settingssidebar_selected_document_menu
- [ ] gutenberg_settingssidebar_selected_block_menu
- [ ] gutenberg_settingssidebar_closed

------------

### **Document Settings Sidebar**

<img width="281" alt="Screen Shot 2019-06-05 at 2 31 52 PM" src="https://user-images.githubusercontent.com/1464705/58992114-c3bc7080-879e-11e9-8ad3-ef61a8f9208a.png">

- [ ] gutenberg_documentsidebar_toggled_panel (pass panel name as prop)
- [ ] gutenberg_documentsidebar_changed_visibility (pass visibility as prop)
- [ ] gutenberg_documentsidebar_changed_publish_date (pass immediately, or scheduled as prop)
- [ ] gutenberg_documentsidebar_changed_post_format (pass post format as prop)
- [ ] gutenberg_documentsidebar_toggled_stick_to_top (pass status as prop)
- [ ] gutenberg_documentsidebar_toggled_pending_review (pass status as prop)
- [ ] gutenberg_documentsidebar_changed_author
- [ ] gutenberg_documentsidebar_selected_trash_post

<img width="142" alt="Screen Shot 2019-06-05 at 2 39 28 PM" src="https://user-images.githubusercontent.com/1464705/58992458-bc499700-879f-11e9-95e0-d7072f189c87.png">

- [ ] gutenberg_documentsidebar_toggled_category
- [ ] gutenberg_documentsidebar_edited_tags

<img width="279" alt="Screen Shot 2019-06-05 at 2 40 08 PM" src="https://user-images.githubusercontent.com/1464705/58992489-d3888480-879f-11e9-8676-5fb684d5899c.png">

- [ ] gutenberg_documentsidebar_selected_set_featured_image
- [ ] gutenberg_documentsidebar_selected_replace_featured_image
- [ ] gutenberg_documentsidebar_selected_remove_featured_image
- [ ] gutenberg_documentsidebar_edited_excerpt
- [ ] gutenberg_documentsidebar_toggled_allow_comments
- [ ] gutenberg_documentsidebar_toggled_allow_pingbacks

------------

### **Block Settings Sidebar**

<img width="144" alt="Screen Shot 2019-06-05 at 3 17 11 PM" src="https://user-images.githubusercontent.com/1464705/58994339-02edc000-87a5-11e9-81e2-177667c6bfe3.png">

Since this sidebar changes depending on the block, I think we need to use more generic events for links, toggles, checkboxes, buttons, and text fields. I'm open to thoughts on this one, here's what I think could work:

_We should pass the provider and block name concatenated as a prop to all of these events._

- [ ] gutenberg_blocksidebar_toggled_panel (pass panel name as prop)
- [ ] gutenberg_blocksidebar_selected_style (pass style name as prop)
- [ ] gutenberg_blocksidebar_toggled_option (pass option name and option status as props)
- [ ] gutenberg_blocksidebar_selected_option (pass option name and option value as props)
- [ ] gutenberg_blocksidebar_edited_field (pass field name and field value as props)
- [ ] gutenberg_blocksidebar_selected_link (pass link text as prop)
- [ ] gutenberg_blocksidebar_edited_custom_css_class

/cc @mtias, @obenland

**Comment by apeatling** [comment]
I know that the crashes tracking is last here, but I think it should be the first thing we add tracking to.

**Comment by apeatling** [comment]
What's missing here, does the naming convention make sense?

**Comment by gwwar** [comment]
> I don't know how to reproduce this (except for these threads, but we definitely need to track it:

This fires for any unexpected JS error, so it'll be a symptom. We should always expect to see some volume of these, but we'll want to monitor to reduce common issues.

Andy noted that he's interested in all event types here for the error report. Stats, Kibana (for the js error stacks) and tracks.

For metrics calls, we can use the common folder of https://github.com/Automattic/wp-calypso/tree/master/apps/wpcom-block-editor to serve scripts to both wp-admin and Calypso. This is served off of widgets.wp.com (Please follow readme steps)

See also PCYsg-l4k-p2

**Comment by apeatling** [comment]
Need to also track when elements are dragged into the editor and a block is inserted that way.

**Comment by obenland** [comment]
> We should pass the provider and block name concatenated as a prop to all of these events.

Could you specify what you mean with `provider`?

**Comment by obenland** [comment]
Updated to 13P after #34037 was estimated at 5.

**Comment by getdave** [comment]
@youknowriad Could we possibly ask your advice here? [This PR](https://github.com/Automattic/wp-calypso/pull/33322) found a way to add tracking to WPCom envs via event delegation on the `document` but I'm not sure whether it's the most sustainable given the number of events Andy wants to track.

We'd appreciate your advice on alternative (and possibly more sustainable) long term solutions.

I've been thinking that perhaps adding [more hooks](https://developer.wordpress.org/block-editor/packages/packages-hooks/) to the Gutenberg Core codebase might be a way forward. We could then hang our tracking off of these whilst also conferring the additional benefit of making Core more extensible to other developers. Is that realistic/sensible from a Core point of view?

Any advice/suggestions would be much appreciated.

**Comment by youknowriad** [comment]
That's challenging!

To be honest, we don't like adding hooks in every possible place in Gutenberg components. Hooks not being a reactive API make it less than ideal for JavaScript extensibility, so we tend to avoid them as much as we can. Also, we need to take into consideration the fact, that any hook added is going to be abused by third-party plugins in all possible ways.

Tracking these events is valuable even from a Core's perspective, I know it's a hard rule to avoid tracking in Core but I don't think that forbids us from building the infrastructure for it without exploiting it.

Imagine a `@wordpress/tracking` package that can be used to gather stats from different packages. By default, it could store in memory and just dump the results as needed: `something like `wp.tracking.getResults` or `wp.data.select( 'core/tracking' ).getResults()` while at the same time offering a unique hook to allow sending this to a custom server.

@mtias do you have thoughts here? what do you think of such approach? Is this acceptable in Core?

**Comment by apeatling** [comment]
> Could you specify what you mean with provider?

An example would be Jetpack or CoBlocks, where is the block coming from?

**Comment by getdave** [comment]
> Imagine a @wordpress/tracking package that can be used to gather stats from different packages. By default, it could store in memory and just dump the results as needed

Thanks @youknowriad. So you're suggesting a package that allows Core to track events at specific points in the code? Then anyone can call the `selectors` that the package exposes and get back the data for a specific event? What I'm struggling with is how we would know _when_ to request the data given that some events are in response to user actions (eg: Block Picker search...etc).

**Comment by obenland** [comment]
> An example would be Jetpack or CoBlocks, where is the block coming from?

We should be able to parse that out of the block namespace

**Comment by youknowriad** [comment]
@getdave I imagine a way to extend the "action" triggered to log an event.

**Comment by getdave** [comment]
@apeatling I've updated the events we need track. However we're not following the event naming conventions (`<source>_<context>_<optional subcontext>_<action>_<optional qualifier>`) so I'm going to amend them. `

**Comment by retrofox** [comment]
> > An example would be Jetpack or CoBlocks, where is the block coming from?

> We should be able to parse that out of the block namespace

AFAIK, the provider is defined as a data property of the block main container element. This data property shapes like `core/paragraph`, `jetpack/recurring-payments`, etc. If I didn't misunderstand, we can pick up the provider splitting it up and picking up the initial part. Is it correct?

Also, I've seen some names that could get confusing like `core-embed/twitter`. I guess in these cases we need to adapt that value with the name to the provider name, in this case, it should be `core`.

I no much idea of how many exceptions we will have, though.

cc @apeatling @obenland

**Comment by obenland** [comment]
>  If I didn't misunderstand, we can pick up the provider splitting it up and picking up the initial part. Is it correct?

That's what I was thinking of, yes. 

I wasn't aware of the `core-embed` namespace. Let's treat it as an exception for now. If there are others that we notice, we can revisit and see if we need a more elaborate solution.

**Comment by apeatling** [comment]
Our base level of metric tracking is sufficient at this point. I am moving the other remaining events to the "Not Doing" pile until a need for them surfaces.

**Comment by getdave** [comment]
@apeatling Can this be closed?

**Comment by davipontesblog** [comment]
I am closing this in favor of the current efforts with core FSE.
@apeatling please reopen if needed!
